// react-responsive-ui/modules/ExpandableMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _class, _temp2;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _ExpandableList = require('./ExpandableList');

var _ExpandableList2 = _interopRequireDefault(_ExpandableList);

var _List = require('./List');

var _List2 = _interopRequireDefault(_List);

var _Close = require('./Close');

var _Close2 = _interopRequireDefault(_Close);

var _focus = require('./utility/focus');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

// `PureComponent` is only available in React >= 15.3.0.
var PureComponent = _react2.default.PureComponent || _react2.default.Component;

var ExpandableMenu = (_temp2 = _class = function (_PureComponent) {
	_inherits(ExpandableMenu, _PureComponent);

	function ExpandableMenu() {
		var _ref;

		var _temp, _this, _ret;

		_classCallCheck(this, ExpandableMenu);

		for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
			args[_key] = arguments[_key];
		}

		return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ExpandableMenu.__proto__ || Object.getPrototypeOf(ExpandableMenu)).call.apply(_ref, [this].concat(args))), _this), _this.onCollapse = function (_ref2) {
			var focusOut = _ref2.focusOut;

			if (!focusOut) {
				_this.focus();
			}
		}, _this.focus = function () {
			return (0, _focus.focus)(_this.toggler);
		}, _this.expand = function () {
			return _this.list.expand();
		}, _this.collapse = function () {
			return _this.list.collapse();
		}, _this.toggle = function () {
			return _this.list.toggle();
		}, _this.storeListRef = function (ref) {
			return _this.list = ref;
		}, _this.storeTogglerRef = function (ref) {
			return _this.toggler = ref;
		}, _this.storeTogglerNode = function (node) {
			return _this.togglerNode = node;
		}, _this.getTogglerNode = function () {
			return _this.togglerNode;
		}, _this.onBlur = function (event) {
			return _this.list && _this.list.onBlur(event);
		}, _this.onClick = function (event) {
			var disabled = _this.props.disabled;


			if (!disabled) {
				_this.toggle();
			}
		}, _this.onKeyDown = function (event) {
			var disabled = _this.props.disabled;


			if (disabled) {
				return;
			}

			if (event.defaultPrevented) {
				return;
			}

			if (event.ctrlKey || event.altKey || event.shiftKey || event.metaKey) {
				return;
			}

			switch (event.keyCode) {
				// "Up" arrow.
				// Select the previous item (if present).
				case 38:
				// "Down" arrow.
				// Select the next item (if present).
				case 40:
					return _this.list.onKeyDown(event);

				// "Enter".
				case 13:
					// Submit containing `<form/>`.
					// Expand otherwise.
					_this.expand();
					return event.preventDefault();
			}
		}, _temp), _possibleConstructorReturn(_this, _ret);
	}

	_createClass(ExpandableMenu, [{
		key: 'render',
		value: function render() {
			var _props = this.props,
			    upward = _props.upward,
			    alignment = _props.alignment,
			    scrollIntoView = _props.scrollIntoView,
			    closeButtonIcon = _props.closeButtonIcon,
			    closeLabel = _props.closeLabel,
			    style = _props.style,
			    className = _props.className,
			    children = _props.children;


			var menuItems = _react2.default.Children.toArray(children);
			var toggler = menuItems.shift();

			return _react2.default.createElement(
				'div',
				{
					style: style,
					className: (0, _classnames2.default)('rrui__menu', className) },
				_react2.default.createElement(
					'div',
					{
						ref: this.storeTogglerNode,
						onBlur: this.onBlur,
						onClick: this.onClick,
						onKeyDown: this.onKeyDown },
					_react2.default.cloneElement(toggler, { ref: this.storeTogglerRef })
				),
				_react2.default.createElement(
					_ExpandableList2.default,
					{
						ref: this.storeListRef,
						upward: upward,
						alignment: alignment,
						scrollIntoView: scrollIntoView,
						scrollMaxItems: 0,
						onCollapse: this.onCollapse,
						onExpand: this.onExpand,
						onTapOutside: this.collapse,
						onFocusOut: this.collapse,
						getTogglerNode: this.getTogglerNode,
						closeButtonIcon: closeButtonIcon,
						closeLabel: closeLabel,
						focusSelectedItem: false,
						shouldCreateButtons: false,
						className: 'rrui__shadow' },
					menuItems
				)
			);
		}
	}]);

	return ExpandableMenu;
}(PureComponent), _class.propTypes = {
	// Component CSS class
	className: _propTypes2.default.string,

	// CSS style object
	style: _propTypes2.default.object,

	// `aria-label` for the "Close" button
	// (which is an "x" visible in fullscreen mode).
	closeLabel: _propTypes2.default.string,

	// The "x" button icon that closes the `<Select/>`
	// in fullscreen mode on mobile devices.
	closeButtonIcon: _propTypes2.default.oneOfType([_propTypes2.default.func, _propTypes2.default.oneOf([false])]).isRequired
}, _class.defaultProps = {
	// The "x" button icon that closes the `<Select/>`
	// in fullscreen mode on mobile devices.
	closeButtonIcon: _Close.CloseIcon
}, _temp2);
exports.default = ExpandableMenu;
