// react-responsive-ui/modules/SlideOutMenu.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.default = undefined;

var _class2, _temp3;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _reactCreateRef = require('react-create-ref');

var _reactCreateRef2 = _interopRequireDefault(_reactCreateRef);

var _PageAndMenu = require('./PageAndMenu');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

// `PureComponent` is only available in React >= 15.3.0.
var PureComponent = _react2.default.PureComponent || _react2.default.Component;

var ContextAwareSlideoutMenu = function (_PureComponent) {
	_inherits(ContextAwareSlideoutMenu, _PureComponent);

	function ContextAwareSlideoutMenu() {
		var _ref;

		var _temp, _this, _ret;

		_classCallCheck(this, ContextAwareSlideoutMenu);

		for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
			args[_key] = arguments[_key];
		}

		return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ContextAwareSlideoutMenu.__proto__ || Object.getPrototypeOf(ContextAwareSlideoutMenu)).call.apply(_ref, [this].concat(args))), _this), _this.slideOutMenu = (0, _reactCreateRef2.default)(), _this.show = function () {
			return _this.slideOutMenu.current.show();
		}, _this.hide = function () {
			return _this.slideOutMenu.current.hide();
		}, _temp), _possibleConstructorReturn(_this, _ret);
	}

	_createClass(ContextAwareSlideoutMenu, [{
		key: 'render',
		value: function render() {
			var _this2 = this;

			return _react2.default.createElement(
				_PageAndMenu.Context.Consumer,
				null,
				function (context) {
					return _react2.default.createElement(SlideoutMenu, _extends({
						ref: _this2.slideOutMenu
					}, _this2.props, {
						registerMenu: context.registerMenu,
						toggleMenu: context.toggleMenu }));
				}
			);
		}
	}]);

	return ContextAwareSlideoutMenu;
}(PureComponent);

// const ContextAwareSlideoutMenu = (props) => (
// 	<Context.Consumer>
// 		{context => (
// 			<SlideoutMenu
// 				{...props}
// 				registerMenu={context.registerMenu}
// 				toggleMenu={context.toggleMenu}/>
// 			)
// 		}
// 	</Context.Consumer>
// )

// export default ContextAwareSlideoutMenu

// Swipeable feature example source code:
// https://github.com/mui-org/material-ui/blob/v1-beta/packages/material-ui/src/SwipeableDrawer/SwipeableDrawer.js

// A slideout menu.


exports.default = ContextAwareSlideoutMenu;
var SlideoutMenu = (_temp3 = _class2 = function (_PureComponent2) {
	_inherits(SlideoutMenu, _PureComponent2);

	function SlideoutMenu() {
		var _ref2;

		var _temp2, _this3, _ret2;

		_classCallCheck(this, SlideoutMenu);

		for (var _len2 = arguments.length, args = Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
			args[_key2] = arguments[_key2];
		}

		return _ret2 = (_temp2 = (_this3 = _possibleConstructorReturn(this, (_ref2 = SlideoutMenu.__proto__ || Object.getPrototypeOf(SlideoutMenu)).call.apply(_ref2, [this].concat(args))), _this3), _this3.state = {
			// show: this.props.isOpen
			show: false
		}, _this3.hide = function () {
			var toggleMenu = _this3.props.toggleMenu;
			var show = _this3.state.show;


			if (show) {
				toggleMenu();
			}
		}, _this3.storeInstance = function (ref) {
			return _this3.menu = ref;
		}, _temp2), _possibleConstructorReturn(_this3, _ret2);
	}

	_createClass(SlideoutMenu, [{
		key: 'componentDidMount',
		value: function componentDidMount() {
			var _this4 = this;

			var registerMenu = this.props.registerMenu;
			var show = this.state.show;


			this.unregister = registerMenu({
				hide: function hide() {
					return _this4.setState({ show: false });
				},
				toggle: function toggle(callback) {
					return _this4.setState(function (state) {
						return { show: !state.show };
					}, callback);
				},
				isShown: function isShown() {
					return _this4.state.show;
				},
				element: function element() {
					return _this4.menu;
				}
			});

			// // Hide on `Back`/`Forward` navigation.
			// window.addEventListener('popstate', this.hide)
		}
	}, {
		key: 'componentWillUnmount',
		value: function componentWillUnmount() {
			this.unregister();

			// window.removeEventListener('popstate', this.hide)
		}

		// componentDidUpdate(prevProps)
		// {
		// 	const { isOpen } = this.props

		// 	if (prevProps.isOpen && !isOpen) {
		// 		this.hide()
		// 	} else if (!prevProps.isOpen && isOpen) {
		// 		this.show()
		// 	}
		// }

	}, {
		key: 'render',
		value: function render() {
			var _props = this.props,
			    anchor = _props.anchor,
			    fullscreen = _props.fullscreen,
			    className = _props.className,
			    style = _props.style,
			    children = _props.children;
			var show = this.state.show;


			return _react2.default.createElement(
				'div',
				{
					ref: this.storeInstance,
					className: (0, _classnames2.default)('rrui__slideout-menu', {
						'rrui__slideout-menu--left': anchor === 'left',
						'rrui__slideout-menu--right': anchor === 'right',
						'rrui__slideout-menu--top': anchor === 'top',
						'rrui__slideout-menu--bottom': anchor === 'bottom',
						'rrui__slideout-menu--fullscreen': fullscreen,
						'rrui__slideout-menu--expanded': show
					}, className),
					style: style },
				children
			);
		}
	}]);

	return SlideoutMenu;
}(PureComponent), _class2.propTypes = {
	anchor: _propTypes2.default.oneOf(['left', 'right', 'top', 'bottom']).isRequired,

	// isOpen : PropTypes.bool.isRequired,

	fullscreen: _propTypes2.default.bool.isRequired,

	toggleMenu: _propTypes2.default.func.isRequired,
	registerMenu: _propTypes2.default.func.isRequired,

	// CSS style object
	style: _propTypes2.default.object,

	// CSS class
	className: _propTypes2.default.string
}, _class2.defaultProps = {
	// isOpen : false,
	anchor: 'left',
	fullscreen: false
}, _temp3);
