// react-responsive-ui/commonjs/TextInputInputComponent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});
exports.default = TextInput;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _TextInputInput = require('./TextInputInput');

var _TextInputInput2 = _interopRequireDefault(_TextInputInput);

var _TextInputLabel = require('./TextInputLabel');

var _TextInputLabel2 = _interopRequireDefault(_TextInputLabel);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// `<input/>` and its `<label/>`.
function TextInput(props) {
	var id = props.id,
	    value = props.value,
	    required = props.required,
	    floatingLabel = props.floatingLabel,
	    label = props.label,
	    placeholder = props.placeholder,
	    multiline = props.multiline,
	    error = props.error,
	    indicateInvalid = props.indicateInvalid,
	    containerRef = props.containerRef,
	    children = props.children;


	return _react2.default.createElement(
		'div',
		{
			ref: containerRef,
			className: (0, _classnames2.default)('rrui__input', {
				'rrui__input--multiline': multiline
			}) },
		_react2.default.createElement(_TextInputInput2.default, props),
		label && _react2.default.createElement(
			_TextInputLabel2.default,
			{
				id: id,
				value: value,
				required: required,
				invalid: indicateInvalid && error,
				floats: floatingLabel && !placeholder },
			label
		),
		children
	);
}

TextInput.propTypes = {
	// Set to `true` to mark the field as required.
	required: _propTypes2.default.bool.isRequired,

	// Labels float by default.
	floatingLabel: _propTypes2.default.bool.isRequired
};

TextInput.defaultProps = {
	// Set to `true` to mark the field as required.
	required: false,

	// Labels float by default.
	floatingLabel: true
};
